File: app/src/main/java/cf/playhi/freezeyou/SettingsUtils.java
package cf.playhi.freezeyou;

import android.app.Activity;
import android.app.admin.DevicePolicyManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.Build;
import android.provider.Settings;

import net.grandcentrix.tray.AppPreferences;

import cf.playhi.freezeyou.utils.DevicePolicyManagerUtils;
import cf.playhi.freezeyou.utils.ServiceUtils;
import cf.playhi.freezeyou.utils.Support;

import static cf.playhi.freezeyou.utils.AccessibilityUtils.isAccessibilitySettingsOn;
import static cf.playhi.freezeyou.utils.AccessibilityUtils.openAccessibilitySettings;
import static cf.playhi.freezeyou.utils.DevicePolicyManagerUtils.getDevicePolicyManager;
import static cf.playhi.freezeyou.utils.DevicePolicyManagerUtils.openDevicePolicyManager;
import static cf.playhi.freezeyou.utils.ToastUtils.showToast;

final class SettingsUtils {
    static void syncAndCheckSharedPreference(Context context, Activity activity,
                                             SharedPreferences sharedPreferences, String s,
                                             AppPreferences appPreferences) {
        switch (s) {
            case "firstIconEnabled":
                if (sharedPreferences.getBoolean(s, true)) {
                    context.getPackageManager().setComponentEnabledSetting(new ComponentName(context, "cf.playhi.freezeyou.FirstIcon"),
                            PackageManager.COMPONENT_ENABLED_STATE_ENABLED, PackageManager.DONT_KILL_APP);
                } else {
                    context.getPackageManager().setComponentEnabledSetting(new ComponentName(context, "cf.playhi.freezeyou.FirstIcon"),
                            PackageManager.COMPONENT_ENABLED_STATE_DISABLED, PackageManager.DONT_KILL_APP);
                }
                showToast(activity, R.string.ciFinishedToast);
                break;
            case "secondIconEnabled":
                if (sharedPreferences.getBoolean(s, true)) {
                    context.getPackageManager().setComponentEnabledSetting(new ComponentName(context, "cf.playhi.freezeyou.SecondIcon"),
                            PackageManager.COMPONENT_ENABLED_STATE_ENABLED, PackageManager.DONT_KILL_APP);
                } else {
                    context.getPackageManager().setComponentEnabledSetting(new ComponentName(context, "cf.playhi.freezeyou.SecondIcon"),
                            PackageManager.COMPONENT_ENABLED_STATE_DISABLED, PackageManager.DONT_KILL_APP);
                }
                showToast(activity, R.string.ciFinishedToast);
                break;
            case "thirdIconEnabled":
                if (sharedPreferences.getBoolean(s, true)) {
                    context.getPackageManager().setComponentEnabledSetting(new ComponentName(context, "cf.playhi.freezeyou.ThirdIcon"),
                            PackageManager.COMPONENT_ENABLED_STATE_ENABLED, PackageManager.DONT_KILL_APP);
                } else {
                    context.getPackageManager().setComponentEnabledSetting(new ComponentName(context, "cf.playhi.freezeyou.ThirdIcon"),
                            PackageManager.COMPONENT_ENABLED_STATE_DISABLED, PackageManager.DONT_KILL_APP);
                }
                showToast(activity, R.string.ciFinishedToast);
                break;
            case "shortCutOneKeyFreezeAdditionalOptions":
                if (!"nothing".equals(sharedPreferences.getString(s, "nothing"))) {
                    appPreferences.put(s, sharedPreferences.getString(s, "nothing"));
                    DevicePolicyManager devicePolicyManager = getDevicePolicyManager(context);
                    if (devicePolicyManager != null && !devicePolicyManager.isAdminActive(
                            new ComponentName(context, DeviceAdminReceiver.class))) {
                        openDevicePolicyManager(context);
                    }
                }
                break;
            case "uiStyleSelection":
                showToast(activity, R.string.willTakeEffectsNextLaunch);
                activity.recreate();
                break;
            case "onekeyFreezeWhenLockScreen":
                appPreferences.put(s, sharedPreferences.getBoolean(s, false));
                if (sharedPreferences.getBoolean(s, false)) {
                    ServiceUtils.startService(
                            context,
                            new Intent(context, ScreenLockOneKeyFreezeService.class));
                } else {
                    context.stopService(new Intent(context, ScreenLockOneKeyFreezeService.class));
                }
                break;
            case "freezeOnceQuit":
            case "avoidFreezeForegroundApplications":
                appPreferences.put(s, sharedPreferences.getBoolean(s, false));
                if (sharedPreferences.getBoolean(s, false) && !isAccessibilitySettingsOn(context)) {
                    showToast(activity, R.string.needActiveAccessibilityService);
                    openAccessibilitySettings(context);
                }
                break;
            case "useForegroundService":// 使用前台服务
            case "openImmediately":// 解冻后立即打开
            case "openAndUFImmediately":// 快捷方式立即解冻并启动
            case "notificationBarDisableSlideOut":// 禁止滑动移除
            case "notificationBarDisableClickDisappear":// 禁止点击消失
            case "lesserToast":// 更少的 Toast
            case "debugModeEnabled":// 启用 调试模式
            case "tryDelApkAfterInstalled"://安装完成后尝试删除安装包
                appPreferences.put(s, sharedPreferences.getBoolean(s, false));
                break;
            case "notificationBarFreezeImmediately":
            case "showInRecents":
            case "notAllowInstallWhenIsObsd":
                appPreferences.put(s, sharedPreferences.getBoolean(s, true));
                break;
            case "organizationName":
                DevicePolicyManagerUtils.checkAndSetOrganizationName(context, sharedPreferences.getString(s, null));
                break;
            case "avoidFreezeNotifyingApplications":
                appPreferences.put(s, sharedPreferences.getBoolean(s, false));
                if (Build.VERSION.SDK_INT >= 21) {//这项不兼容 5.0 以下了
                    String enabledNotificationListeners = Settings.Secure.getString(context.getContentResolver(), "enabled_notification_listeners");
                    if (enabledNotificationListeners != null && !enabledNotificationListeners.contains("cf." + "playhi." + "freezeyou")) {
                        try {
                            context.startActivity(new Intent("android.settings.ACTION_NOTIFICATION_LISTENER_SETTINGS"));
                        } catch (Exception e) {
                            showToast(activity, R.string.failed);
                        }
                    }
                }
                break;
            case "enableInstallPkgFunc":
                if (sharedPreferences.getBoolean(s, true)) {
                    context.getPackageManager().setComponentEnabledSetting(new ComponentName(context, "cf.playhi.freezeyou.InstallPackagesActivity"),
                            PackageManager.COMPONENT_ENABLED_STATE_ENABLED, PackageManager.DONT_KILL_APP);
                } else {
                    context.getPackageManager().setComponentEnabledSetting(new ComponentName(context, "cf.playhi.freezeyou.InstallPackagesActivity"),
                            PackageManager.COMPONENT_ENABLED_STATE_DISABLED, PackageManager.DONT_KILL_APP);
                }
                break;
            case "languagePref":
                Support.checkLanguage(activity);
                showToast(activity, R.string.willTakeEffectsNextLaunch);
                break;
            default:
                break;
        }
    }
}

File: app/src/main/java/cf/playhi/freezeyou/utils/SettingsUtils.kt
package cf.playhi.freezeyou.utils

import android.app.Activity
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.os.Build
import android.provider.Settings
import cf.playhi.freezeyou.R
import cf.playhi.freezeyou.fuf.FUFSinglePackage.Companion.API_FREEZEYOU_LEGACY_AUTO
import cf.playhi.freezeyou.fuf.FUFSinglePackage.Companion.API_FREEZEYOU_MROOT_DPM
import cf.playhi.freezeyou.fuf.FUFSinglePackage.Companion.API_FREEZEYOU_MROOT_PROFILE_OWNER
import cf.playhi.freezeyou.fuf.FUFSinglePackage.Companion.API_FREEZEYOU_ROOT_DISABLE_ENABLE
import cf.playhi.freezeyou.fuf.FUFSinglePackage.Companion.API_FREEZEYOU_ROOT_UNHIDE_HIDE
import cf.playhi.freezeyou.fuf.FUFSinglePackage.Companion.API_FREEZEYOU_SYSTEM_APP_ENABLE_DISABLE
import cf.playhi.freezeyou.fuf.FUFSinglePackage.Companion.API_FREEZEYOU_SYSTEM_APP_ENABLE_DISABLE_UNTIL_USED
import cf.playhi.freezeyou.fuf.FUFSinglePackage.Companion.API_FREEZEYOU_SYSTEM_APP_ENABLE_DISABLE_USER
import cf.playhi.freezeyou.service.ScreenLockOneKeyFreezeService
import cf.playhi.freezeyou.storage.key.*
import cf.playhi.freezeyou.storage.key.DefaultMultiProcessMMKVStorageBooleanKeys.*
import cf.playhi.freezeyou.storage.key.DefaultMultiProcessMMKVStorageStringKeys.*
import cf.playhi.freezeyou.storage.key.DefaultSharedPreferenceStorageBooleanKeys.enableInstallPkgFunc
import cf.playhi.freezeyou.storage.key.DefaultSharedPreferenceStorageStringKeys.mainActivityPattern
import cf.playhi.freezeyou.storage.key.DefaultSharedPreferenceStorageStringKeys.organizationName
import cf.playhi.freezeyou.utils.ToastUtils.showToast

object SettingsUtils {

    /**
     * Show toast, check availability, sync to MMKV, etc.
     */
    @JvmStatic
    fun checkPreferenceData(
        context: Context, activity: Activity,
        sharedPreferences: SharedPreferences?, key: String?
    ) {
        if (key == null || sharedPreferences == null) return

        val abstractKey = convertToAbstractKey(key) ?: return

        syncMMKVDataToMMKVWhenSharedPreferenceDataChanged(context, sharedPreferences, abstractKey)

        when (abstractKey) {
            uiStyleSelection,
            allowFollowSystemAutoSwitchDarkMode,
            mainActivityPattern,
            languagePref ->
                showToast(
                    activity,
                    R.string.willTakeEffectsNextLaunch
                )
            onekeyFreezeWhenLockScreen -> {
                if (onekeyFreezeWhenLockScreen.getValue()) {
                    ServiceUtils.startService(
                        context,
                        Intent(context, ScreenLockOneKeyFreezeService::class.java)
                    )
                } else {
                    context.stopService(Intent(context, ScreenLockOneKeyFreezeService::class.java))
                }
            }
            freezeOnceQuit,
            avoidFreezeForegroundApplications,
            tryToAvoidUpdateWhenUsing -> {
                if (abstractKey.getValue() as Boolean
                    && !AccessibilityUtils.isAccessibilitySettingsOn(context)
                ) {
                    showToast(activity, R.string.needActiveAccessibilityService)
                    AccessibilityUtils.openAccessibilitySettings(context)
                }
            }
            organizationName ->
                DevicePolicyManagerUtils.checkAndSetOrganizationName(
                    context,
                    sharedPreferences.getString(key, null)
                )
            avoidFreezeNotifyingApplications -> {
                if (Build.VERSION.SDK_INT >= 21) {
                    val enabledNotificationListeners = Settings.Secure.getString(
                        context.contentResolver, "enabled_notification_listeners"
                    )
                    if (enabledNotificationListeners != null
                        && !enabledNotificationListeners.contains("cf." + "playhi." + "freezeyou")
                    ) {
                        try {
                            activity.startActivity(
                                Intent("android.settings.ACTION_NOTIFICATION_LISTENER_SETTINGS")
                            )
                        } catch (e: Exception) {
                            showToast(activity, R.string.failed)
                        }
                    }
                }
            }
            enableInstallPkgFunc ->
                if (sharedPreferences.getBoolean(key, enableInstallPkgFunc.defaultValue())) {
                    context.packageManager.setComponentEnabledSetting(
                        ComponentName(context, "cf.playhi.freezeyou.InstallPackagesActivity"),
                        PackageManager.COMPONENT_ENABLED_STATE_ENABLED,
                        PackageManager.DONT_KILL_APP
                    )
                } else {
                    context.packageManager.setComponentEnabledSetting(
                        ComponentName(context, "cf.playhi.freezeyou.InstallPackagesActivity"),
                        PackageManager.COMPONENT_ENABLED_STATE_DISABLED,
                        PackageManager.DONT_KILL_APP
                    )
                }
            selectFUFMode -> {
                when (selectFUFMode.getValue()?.toInt()) {
                    API_FREEZEYOU_MROOT_DPM ->
                        if (!DevicePolicyManagerUtils.isDeviceOwner(context)) {
                            showToast(context, R.string.noMRootPermission)
                        }
                    API_FREEZEYOU_MROOT_PROFILE_OWNER ->
                        if (!DevicePolicyManagerUtils.isProfileOwner(context)) {
                            showToast(context, R.string.isNotProfileOwner)
                        }
                    API_FREEZEYOU_ROOT_DISABLE_ENABLE,
                    API_FREEZEYOU_ROOT_UNHIDE_HIDE ->
                        if (!FUFUtils.checkRootPermission()) {
                            showToast(context, R.string.noRootPermission)
                        }
                    @Suppress("DEPRECATION")
                    API_FREEZEYOU_LEGACY_AUTO ->
                        if (!(FUFUtils.checkRootPermission()
                                    || DevicePolicyManagerUtils.isDeviceOwner(context))
                        ) {
                            showToast(context, R.string.insufficientPermission)
                        }
                    API_FREEZEYOU_SYSTEM_APP_ENABLE_DISABLE_UNTIL_USED,
                    API_FREEZEYOU_SYSTEM_APP_ENABLE_DISABLE_USER,
                    API_FREEZEYOU_SYSTEM_APP_ENABLE_DISABLE ->
                        if (!FUFUtils.isSystemApp(context)) {
                            showToast(context, R.string.insufficientPermission)
                        }
                    else -> showToast(context, R.string.unknown)
                }
            }
        }
    }

    private fun syncMMKVDataToMMKVWhenSharedPreferenceDataChanged(
        context: Context,
        sharedPreferences: SharedPreferences,
        key: AbstractKey<*>
    ) {
        when (key) {
            is DefaultMultiProcessMMKVStorageBooleanKeys -> {
                key.run {
                    setValue(context, sharedPreferences.getBoolean(key.name, key.defaultValue()))
                    sync()
                }
            }
            is DefaultMultiProcessMMKVStorageStringKeys -> {
                key.run {
                    setValue(context, sharedPreferences.getString(key.name, key.defaultValue()))
                    sync()
                }
            }
        }
    }

    fun changeIconEntryComponentState(context: Context, newValue: Boolean, cls: String) {
        if (newValue) {
            context.packageManager.setComponentEnabledSetting(
                ComponentName(context, cls),
                PackageManager.COMPONENT_ENABLED_STATE_ENABLED,
                PackageManager.DONT_KILL_APP
            )
        } else {
            context.packageManager.setComponentEnabledSetting(
                ComponentName(context, cls),
                PackageManager.COMPONENT_ENABLED_STATE_DISABLED,
                PackageManager.DONT_KILL_APP
            )
        }
        showToast(context, R.string.ciFinishedToast)
    }

    private fun convertToAbstractKey(key: String): AbstractKey<*>? {
        try {
            return DefaultMultiProcessMMKVStorageBooleanKeys.valueOf(key)
        } catch (_: IllegalArgumentException) {
        }

        try {
            return DefaultSharedPreferenceStorageBooleanKeys.valueOf(key)
        } catch (_: IllegalArgumentException) {
        }

        try {
            return DefaultMultiProcessMMKVStorageStringKeys.valueOf(key)
        } catch (_: IllegalArgumentException) {
        }

        try {
            return DefaultSharedPreferenceStorageStringKeys.valueOf(key)
        } catch (_: IllegalArgumentException) {
        }

        return null
    }
}